File: sdk/npm/src/providers/tmail-link.ts
/**
 * tmail-link 渠道 - tmail.link (Django)
 *
 * 创建邮箱: GET / → 从 HTML 提取随机分配的邮箱地址
 * 获取邮件:
 *   1. GET /inbox/{email}/ → 获取 csrftoken cookie
 *   2. POST /inbox/{email}/ → form-data(format=json, csrfmiddlewaretoken) → JSON 邮件列表
 *   3. GET /inbox/{email}/{key}/ → 单封邮件详情 HTML
 *
 * token 存储: csrftoken cookie 值
 */

import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "tmail-link";
const BASE = "https://tmail.link";

const DEFAULT_HEADERS: Record<string, string> = {
  Accept:
    "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
  "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
  "Cache-Control": "no-cache",
  DNT: "1",
  Pragma: "no-cache",
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36",
};

/** 匹配 data-default="xxx@tmail.link" 属性 */
const DATA_DEFAULT_RE = /data-default="([^"]+@tmail\.link)"/i;
/** 通用匹配 xxx@tmail.link 邮箱 */
const EMAIL_RE = /\b([a-zA-Z0-9._%+-]+@tmail\.link)\b/;

/** 从 Set-Cookie 头中提取 csrftoken 值 */
function extractCsrfToken(headers: Headers): string {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  const lines: string[] =
    typeof h.getSetCookie === "function"
      ? h.getSetCookie()
      : headers.get("set-cookie")
        ? [headers.get("set-cookie")!]
        : [];

  for (const line of lines) {
    const match = /csrftoken=([^;]+)/i.exec(line);
    if (match?.[1]) return match[1].trim();
  }
  return "";
}

/**
 * 创建临时邮箱
 * 1. GET https://tmail.link/ → 从 HTML 中提取分配的邮箱地址
 * 2. GET https://tmail.link/inbox/{email}/ → 获取 csrftoken cookie
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const res = await fetchWithTimeout(`${BASE}/`, {
    method: "GET",
    headers: DEFAULT_HEADERS,
  });
  if (!res.ok) {
    throw new Error(`tmail-link: 创建邮箱失败 HTTP ${res.status}`);
  }

  const html = await res.text();

  /* 提取邮箱地址: 优先匹配 data-default，回退到通用正则 */
  let email = "";
  const dm = DATA_DEFAULT_RE.exec(html);
  if (dm?.[1]) {
    email = dm[1].trim();
  } else {
    const em = EMAIL_RE.exec(html);
    if (em?.[1]) {
      email = em[1].trim();
    }
  }
  if (!email) {
    throw new Error("tmail-link: 未从页面提取到邮箱地址");
  }

  /* 请求 inbox 页面以获取 csrftoken cookie（首页不设置 cookie） */
  const inboxRes = await fetchWithTimeout(
    `${BASE}/inbox/${encodeURIComponent(email)}/`,
    {
      method: "GET",
      headers: DEFAULT_HEADERS,
    },
  );
  if (!inboxRes.ok) {
    throw new Error(`tmail-link: 获取收件箱页面失败 HTTP ${inboxRes.status}`);
  }
  await inboxRes.text();

  const csrftoken = extractCsrfToken(inboxRes.headers);
  if (!csrftoken) {
    throw new Error("tmail-link: 未获取到 csrftoken");
  }

  return {
    channel: CHANNEL,
    email,
    token: csrftoken,
  };
}

/**
 * 获取邮件列表
 *
 * 1. GET /inbox/{email}/ → 刷新 csrftoken cookie
 * 2. POST /inbox/{email}/ → 带 csrfmiddlewaretoken 请求 JSON 邮件列表
 * 3. 对每封邮件 GET /inbox/{email}/{key}/ 获取详情
 *
 * @param token - csrftoken cookie 值
 * @param email - 邮箱地址
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  if (!email?.trim()) {
    throw new Error("tmail-link: 邮箱地址为空");
  }

  const inboxUrl = `${BASE}/inbox/${encodeURIComponent(email)}/`;

  /* 步骤1: GET inbox 页面，获取最新 csrftoken */
  const getRes = await fetchWithTimeout(inboxUrl, {
    method: "GET",
    headers: {
      ...DEFAULT_HEADERS,
      Cookie: `csrftoken=${token}`,
    },
  });
  if (!getRes.ok) {
    throw new Error(`tmail-link: 获取收件箱页面失败 HTTP ${getRes.status}`);
  }

  /* 更新 csrftoken（如果服务端刷新了 cookie） */
  const freshToken = extractCsrfToken(getRes.headers) || token;

  /* 步骤2: POST inbox 页面请求 JSON 格式邮件列表 */
  const formBody = new URLSearchParams();
  formBody.set("format", "json");
  formBody.set("csrfmiddlewaretoken", freshToken);

  const postRes = await fetchWithTimeout(inboxUrl, {
    method: "POST",
    headers: {
      ...DEFAULT_HEADERS,
      "Content-Type": "application/x-www-form-urlencoded",
      Cookie: `csrftoken=${freshToken}`,
      "X-CSRFToken": freshToken,
      Referer: inboxUrl,
    },
    body: formBody.toString(),
  });
  if (!postRes.ok) {
    throw new Error(`tmail-link: 获取邮件列表失败 HTTP ${postRes.status}`);
  }

  const data = (await postRes.json()) as { messages?: any[] };
  const messages = Array.isArray(data.messages) ? data.messages : [];

  /* 步骤3: 对每封邮件获取详情页 HTML */
  const out: Email[] = [];
  for (const msg of messages) {
    const key = msg.key || msg.id || "";
    let html = "";
    let text = "";

    if (key) {
      try {
        const detailUrl = `${BASE}/inbox/${encodeURIComponent(email)}/${encodeURIComponent(key)}/`;
        const dr = await fetchWithTimeout(detailUrl, {
          method: "GET",
          headers: {
            ...DEFAULT_HEADERS,
            Cookie: `csrftoken=${freshToken}`,
            Referer: inboxUrl,
          },
        });
        if (dr.ok) {
          html = await dr.text();
        }
      } catch {
        /* 获取详情失败不影响列表返回 */
      }
    }

    out.push(
      normalizeEmail(
        {
          id: String(key),
          from: msg.sender || msg.from || "",
          to: email,
          subject: msg.subject || "",
          text,
          html,
          date: msg.date || msg.created_at || "",
        },
        email,
      ),
    );
  }

  return out;
}

File: sdk/go/provider/tmail_link.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"regexp"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

/**
 * TmailLink — https://tmail.link
 * Django 临时邮箱服务（CSRF token 认证）
 * 流程:
 *   1. GET / → HTML 包含随机邮箱地址 (xxx@tmail.link)
 *   2. GET /inbox/{email}/ → 获取 csrftoken cookie
 *   3. POST /inbox/{email}/ + format=json + csrfmiddlewaretoken → JSON 邮件列表
 * Token 格式: csrftoken 字符串
 */

const tmailLinkBase = "https://tmail.link"

var tmailLinkEmailRegex = regexp.MustCompile(`([a-zA-Z0-9._%+-]+@tmail\.link)`)

/**
 * TmailLinkGenerate — 创建 tmail.link 临时邮箱
 */
func TmailLinkGenerate() (*CreatedMailbox, error) {
	client := HTTPClient()

	/* 步骤1: GET / 获取随机邮箱地址 */
	req, err := http.NewRequest("GET", tmailLinkBase+"/", nil)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: 创建请求失败: %w", err)
	}
	req.Header.Set("Accept", "text/html")
	req.Header.Set("User-Agent", getCurrentUA())

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: 请求失败: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: 读取响应失败: %w", err)
	}

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("tmail-link: HTTP %d", resp.StatusCode)
	}

	matches := tmailLinkEmailRegex.FindSubmatch(body)
	if matches == nil || len(matches) < 2 {
		return nil, fmt.Errorf("tmail-link: 无法从响应中解析邮箱地址")
	}
	email := string(matches[1])

	/* 步骤2: GET /inbox/{email}/ 获取 csrftoken cookie */
	inboxURL := fmt.Sprintf("%s/inbox/%s/", tmailLinkBase, url.PathEscape(email))
	req2, err := http.NewRequest("GET", inboxURL, nil)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: 创建 inbox 请求失败: %w", err)
	}
	req2.Header.Set("Accept", "text/html")
	req2.Header.Set("User-Agent", getCurrentUA())

	resp2, err := client.Do(req2)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: inbox 请求失败: %w", err)
	}
	io.ReadAll(resp2.Body)
	resp2.Body.Close()

	var csrfToken string
	for _, cookie := range resp2.Cookies() {
		if cookie.Name == "csrftoken" {
			csrfToken = cookie.Value
			break
		}
	}
	if csrfToken == "" {
		return nil, fmt.Errorf("tmail-link: 未获取到 csrftoken")
	}

	return &CreatedMailbox{
		Channel: "tmail-link",
		Email:   email,
		Token:   csrfToken,
	}, nil
}

/**
 * TmailLinkGetEmails — 获取 tmail.link 邮件列表
 */
func TmailLinkGetEmails(token, email string) ([]NormEmail, error) {
	if token == "" {
		return nil, fmt.Errorf("tmail-link: token 为空")
	}
	if email == "" {
		return nil, fmt.Errorf("tmail-link: 邮箱地址为空")
	}

	client := HTTPClient()
	inboxURL := fmt.Sprintf("%s/inbox/%s/", tmailLinkBase, url.PathEscape(email))

	/* 步骤1: GET inbox 页面获取最新 csrftoken */
	req, err := http.NewRequest("GET", inboxURL, nil)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: 创建 GET inbox 请求失败: %w", err)
	}
	req.Header.Set("Accept", "text/html")
	req.Header.Set("User-Agent", getCurrentUA())
	req.Header.Set("Cookie", "csrftoken="+token)

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: GET inbox 请求失败: %w", err)
	}
	io.ReadAll(resp.Body)
	resp.Body.Close()

	freshToken := token
	for _, cookie := range resp.Cookies() {
		if cookie.Name == "csrftoken" && cookie.Value != "" {
			freshToken = cookie.Value
			break
		}
	}

	/* 步骤2: POST inbox 获取 JSON 邮件列表 */
	formData := url.Values{}
	formData.Set("format", "json")
	formData.Set("csrfmiddlewaretoken", freshToken)

	req2, err := http.NewRequest("POST", inboxURL, strings.NewReader(formData.Encode()))
	if err != nil {
		return nil, fmt.Errorf("tmail-link: 创建 POST inbox 请求失败: %w", err)
	}
	req2.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req2.Header.Set("Accept", "application/json")
	req2.Header.Set("User-Agent", getCurrentUA())
	req2.Header.Set("Cookie", "csrftoken="+freshToken)
	req2.Header.Set("X-CSRFToken", freshToken)
	req2.Header.Set("Referer", inboxURL)

	resp2, err := client.Do(req2)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: POST inbox 请求失败: %w", err)
	}
	defer resp2.Body.Close()

	if resp2.StatusCode < 200 || resp2.StatusCode >= 300 {
		return nil, fmt.Errorf("tmail-link: 邮件请求 HTTP %d", resp2.StatusCode)
	}

	body, err := io.ReadAll(resp2.Body)
	if err != nil {
		return nil, fmt.Errorf("tmail-link: 读取邮件响应失败: %w", err)
	}

	var data struct {
		Messages []json.RawMessage `json:"messages"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("tmail-link: 邮件响应解析失败: %w", err)
	}

	if len(data.Messages) == 0 {
		return []NormEmail{}, nil
	}

	out := make([]NormEmail, 0, len(data.Messages))
	for _, raw := range data.Messages {
		var msg map[string]interface{}
		if err := json.Unmarshal(raw, &msg); err != nil {
			continue
		}
		row := map[string]any{
			"id":         msg["key"],
			"from":       firstStr(msg, "sender", "from"),
			"to":         email,
			"subject":    msg["subject"],
			"text":       firstStr(msg, "body", "text"),
			"html":       firstStr(msg, "html", "body"),
			"created_at": firstStr(msg, "date", "created_at"),
		}
		out = append(out, NormalizeMap(row, email))
	}
	return out, nil
}
